`improve`: write modified config values back to the maker `config.toml` file

wthrajat opened the pull request:
```rs
        let mut config = MakerConfig::new(Some(&data_dir.join("config.toml")))?;

        if let Some(port) = port {
            config.port = port;
        }

        if let Some(rpc_port) = rpc_port {
            config.rpc_port = rpc_port;
        }

        if let Some(socks_port) = socks_port {
            config.socks_port = socks_port;
        }

        if let Some(connection_type) = connection_type {
            config.connection_type = connection_type;
        }
```

Maker config parameters like `port`, `rpc_port`, `socks_port`, `connection_type` when updated with new values, does NOT reflect in the `config.toml` file.
This PR aims to fix that.

wthrajat commented:
Clippy has been recently updated with a new lint: `Too long first doc paragraph` as described in [doc_mod.rs](https://github.com/rust-lang/rust-clippy/blob/master/clippy_lints/src/doc/mod.rs#L454).
Pushed the fix in 7d3ff35

mojoX911 commented:
@wthrajat can you make the long doc fix in a separate PR and I will merge it quickly. This error will trickle into other PRs too.

mojoX911 commented on src/maker/api.rs:
```diff
@@ -203,7 +203,18 @@ impl Maker {
 
-        // TODO: Write the modified config back to the file.
+        // Writing the new modified config back to the `config.toml` file:
+        let updated_config = format!(
+            r#"
+            port = {}
+            rpc_port = {}
+            socks_port = {}
+            connection_type = {:?}
+            "#,
+            config.port, config.rpc_port, config.socks_port, config.connection_type
+        );
+        std::fs::write(data_dir.join("config.toml"), updated_config)
+            .expect("Error while writing the configuration to the config.toml!");
```
Good catch. Don't we need it for Taker too?

Shourya742 commented:
@wthrajat Can you squeeze the commits.

mojoX911 commented on src/taker/api.rs:
```diff
@@ -222,6 +222,16 @@ impl Taker {
+            "#,
+            config.connection_type,
+        );
+        std::fs::write(data_dir.join("config.toml"), updated_config)
+            .expect("Error while updating the configuration to the config.toml!");
```
Wait. This will replace the entire content of the file with just this one parameter.

mojoX911 commented on src/maker/config.rs:
```diff
@@ -193,6 +193,50 @@ impl MakerConfig {
     }
+
+    // Method to manually serialize the Maker Config into a TOML string
+    pub fn to_toml_string(&self) -> String {
+        format!(
+            r#"
+        port = {}
```
Just the Toml string isn't very useful as for all cases we immediately gonna write it in a file. So lets just put that fs::write line also in the API. Will simplify the calling sites.

mojoX911 replied:
Unless you can think of some use case where we will get the toml string but not write the file.

mojoX911 commented on src/maker/config.rs:
```diff
@@ -195,8 +195,8 @@ impl MakerConfig {
     // Method to manually serialize the Maker Config into a TOML string
-    pub fn to_toml_string(&self) -> String {
-        format!(
+    pub fn update_maker_config(&self, file_path: &std::path::Path) {
```
Call this `write_to_file`. Because we can use this even when writing fresh config to a file. Not always update.

mojoX911 reviewed (commented):
Ack.

Now that we have a write to file API, we don't need the separate write_default_config function. We can just create the default config and call `write_to_file`.

mojoX911 commented:
@wthrajat any update on this?

wthrajat commented:
@mojoX911 
- **Key Changes**:
  - Removed the `write_default_config` function from `utill.rs`.
  - Introduced `update_directory_config` for the directory server.
  - Updated `MakerConfig` and `TakerConfig` to use their respective `write_to_file` methods.

- **Reasoning**:
  - The `write_default_config` function was being used in multiple places, including `directory.rs`. Since we cannot reuse `write_to_file` from `MakerConfig` or `TakerConfig` for the directory server (unless I implement a common trait), we introduced `update_directory_config` specifically for the directory server.

- **Testing**:
  - Updated some unit tests to ensure the changes work as expected.
  - All tests are passing.

Shourya742 commented on src/maker/config.rs:
```diff
@@ -193,6 +194,17 @@ impl MakerConfig {
             .unwrap_or(default_config.connection_type),
         })
     }
+
+    // Method to write the provided TOML string to a file
+    pub fn write_to_file(self, path: &PathBuf, toml_data: String) -> std::io::Result<()> {
+        if let Some(parent) = path.parent() {
+            std::fs::create_dir_all(parent)?;
+        }
+        let mut file = std::fs::File::create(path)?;
+        file.write_all(toml_data.as_bytes())?;
+        file.flush()?;
+        Ok(())
+    }
```
Since writing to a file is now handled by a method, the toml_data string can be created within the method itself, rather than being generated at the call site and passed in.

Shourya742 reviewed (changes requested):
I didn't quite understand the reasoning behind removing the `write_default_config` function, which simply writes data to the descriptor buffer and then flushes it to persistent storage. We could just use the function from the utility and pass the `toml_data` string. Could you explain what the issue was with this approach?

mojoX911 commented on src/market/directory.rs:
```diff
@@ -121,6 +121,19 @@ impl DirectoryServer {
             addresses,
         })
     }
+    pub fn update_directory_config(
+        &self,
+        path: &PathBuf,
+        toml_data: String,
+    ) -> std::io::Result<()> {
+        if let Some(parent) = path.parent() {
+            fs::create_dir_all(parent)?;
+        }
+        let mut file = std::fs::File::create(path)?;
+        file.write_all(toml_data.as_bytes())?;
+        file.flush()?;
+        Ok(())
+    }
```
Remove &self. Make this an independent util function. 
It could even be a generic util function `write_toml_to_file`. That can be useful later also in other modules.

mojoX911 commented on src/taker/config.rs:
```diff
@@ -166,31 +165,59 @@ impl TakerConfig {
-                        rpc_port = 8081\n
-                        ",
-    );
-    write_default_config(config_path, config_string).unwrap();
+    let config = TakerConfig::default();
+    config.write_to_file(config_path).unwrap();
 }
```
2 line method. can be removed.

mojoX911 commented on src/utill.rs:
```diff
@@ -350,17 +350,6 @@ pub fn parse_field<T: std::str::FromStr>(value: Option<&String>, default: T) ->
     }
 }
 
-/// Function to write data to default toml files
-pub fn write_default_config(path: &PathBuf, toml_data: String) -> std::io::Result<()> {
-    if let Some(parent) = path.parent() {
-        fs::create_dir_all(parent)?;
-    }
-    let mut file = File::create(path)?;
-    file.write_all(toml_data.as_bytes())?;
-    file.flush()?;
-    Ok(())
-}
-
```
Change this to `write_toml_to_file` and use that in the directory server. 

Also directly call `create_dir_all`. `if let` is not required here.

mojoX911 commented on src/taker/api.rs:
```diff
@@ -222,6 +222,11 @@ impl Taker {
             config.connection_type = connection_type;
         }
 
+        // Update the config file
+        config
+            .clone()
+            .write_to_file(&data_dir.join("config.toml"))?;
+
```
redundant clone and comment.

mojoX911 commented on src/taker/config.rs:
```diff
@@ -166,31 +165,59 @@ impl TakerConfig {
+            self.connection_type,
+            self.rpc_port
+        );
+        if let Some(parent) = path.parent() {
+            std::fs::create_dir_all(parent)?;
+        }
```
`if let` not required.

Shourya742 reviewed (changes requested):
We are almost there. Could you please add the same logic for the directory server? I noticed that we are still using the method from the utility. Everything else looks good. Additionally, if possible, could you organize the commit history?

wthrajat commented:
@Shourya742: Regarding:
> I noticed that we are still using the method from the utility.

[Here](https://github.com/citadel-tech/coinswap/pull/246#discussion_r1807658372) we have concluded to use the utill method only, renaming it as write_toml_to_file()

Shourya742 commented:
> @Shourya742: Regarding:
> 
> > I noticed that we are still using the method from the utility.
> 
> [Here](https://github.com/citadel-tech/coinswap/pull/246#discussion_r1807658372) we have concluded to use the utill method only, renaming it as `write_toml_to_file()`

What I'm suggesting is to associate a `write_to_toml` method with the directory config. This would make our implementation consistent across all config structs, eliminating the need for a separate TOML utility, as I don't see any other use case for it.

mojoX911 commented:
I would its okay for now, and add that when the need arises. It will arise when a separate the crates. But its very far away in the roadmap. In the meantime I would prefer to reduce duplicate code. So a single util method for toml writing makes sense. And who knows, by that far in future, we might find more usecase for it.  

It will also not be just the toml utility. Many other things will need to be localised. We will have to refactor through all of that. Lets not burden ourselves now on stuffs that far away in future. We do it when we do it. Till then better to keep architecture as per single crate design.

mojoX911 commented:
Oh wait. We are already duplicating the code. We should just use the `write_to_toml` util everywhere. Or makes sense to just remove it and inline the function. Its just few lines.

mojoX911 commented on src/maker/config.rs:
```diff
@@ -193,35 +194,61 @@ impl MakerConfig {
+        );
 
-    write_default_config(config_path, config_string)?;
+        std::fs::create_dir_all(path.parent().expect("Path should NOT be root!"))?;
+        let mut file = std::fs::File::create(path)?;
+        file.write_all(toml_data.as_bytes())?;
+        file.flush()?;
+        Ok(())
```
Either use `write_toml_to_file` everywhere, or inline the function like this in `directory.rs` also. 

Upto you. I don't have strong opinion on either.

wthrajat commented:
@Shourya742, @mojoX911:

> Upto you. I don't have strong opinion on either.

I think I will go with inlining it in the `directory.rs` file -- which means removing the `write_toml_to_file()` completely since it was only being used there.

Reason being I have a few ideas for writing the configs of `maker, taker` and `directory` as I am thinking of implementing a trait (but in another PR) for writing configs that can be used in all three structs (`MakerConfig`, `TakerConfig` and `DirectoryServer`) and more structs if needed. That will reduce a lot of code.

mojoX911 reviewed (approved):
Ack..

Shourya742 reviewed (approved):
ACK. Just squash the commits
